Decode transaction from hex

natsoni opened the pull request:
Fixes #5629

This PR adds client-side functionality for decoding transactions at `tx/preview`:

https://github.com/user-attachments/assets/98a6e79e-6767-43ab-9379-fa5cb0db213c

There is an option to disable fetching of input prevouts ("offline mode"), in this case some transaction data is missing and a message is displayed: 

https://github.com/user-attachments/assets/8acf4bdb-665e-4fd1-8685-db5a6b960bc7

To get the decoded transaction's prevouts, we query the full parent transaction(s). This is suboptimal because we don't actually need the full transaction data to decode the transaction, but it comes with benefits compared to using `gettxout` command:

- full parent transaction data (notably its weight and fee) can be used for estimating the effective fee rate of the decoded transaction before broadcasting (not implemented yet)
- using `gettxout` to fetch an output spent by an unconfirmed transaction would need two calls (the second specifying `include_mempool=false`)
- this approach can significantly reduce the number of requests compared to `gettxout`, especially when many inputs originate from the same parent transaction

Of course this choice can be costly if we try to decode a transaction with a lot of inputs from different parents. To mitigate this we limit the number of transactions we can fetch to 20: 

<img width="600" alt="Screenshot 2024-11-27 at 21 02 56" src="https://github.com/user-attachments/assets/876dc31d-65f9-4ea0-8f11-08d93b4a5778">

A good compromise to fix this issue in the future would be to use `gettxout` by default and fallback to fetching full parent transactions if we need them.

mononaut commented:
this looks awesome, great job!

I'll review properly later, but a couple of quick comments/thoughts:

(1) could you please annotate any borrowed code with a link to the original source? apart from anything else it's helpful to know where stuff came from during review to understand the context & design choices.

(2) I think we should add a backend API endpoint for the prevouts instead of trying to do it all client-side.

looking up unconfirmed prevouts (including total vsize etc for the future CPFP feature) is essentially free for the backend since it's all conveniently indexed in memory. then it's just a matter of fetching txos for any confirmed inputs, where we only care about the value and scriptpubkey and don't need the full parent transactions.

I can add `mempool/electrs` support for batched prevout requests if necessary to accelerate that in production.

(3) does this work with *unsigned* txs? if so, how do you handle vsizes & fee rates etc?

(4) it would be neat to support PSBTs. that's definitely a job for a future PR, but just mentioning it now in case it makes a difference to the design.

(5) offline mode is an excellent idea.

natsoni commented:
> **(1)** could you please annotate any borrowed code with a link to the original source? apart from anything else it's helpful to know where stuff came from during review to understand the context & design choices.

Sure, will do! some of the code for decoding raw transactions has been written by you in the mempool backend. I also adapted some logic from the bitcoinjs repo.

> **(2)** I think we should add a backend API endpoint for the prevouts instead of trying to do it all client-side.
> 
> looking up unconfirmed prevouts (including total vsize etc for the future CPFP feature) is essentially free for the backend since it's all conveniently indexed in memory. then it's just a matter of fetching txos for any confirmed inputs, where we only care about the value and scriptpubkey and don't need the full parent transactions.
> 
> I can add `mempool/electrs` support for batched prevout requests if necessary to accelerate that in production.

I agree that an endpoint for directly fetching prevouts instead of handling them client-side would be nice. And a batched prevout endpoint in `mempool/electrs` would be neat! 
The process for querying a prevout would be: 

1. See if the prevout is unconfirmed in-memory
2. If not, call `gettxout` with `include_mempool=false`

IIUC this should always return the prevout if the corresponding output exists, and also provide extra data for CPFP calculation if it's unconfirmed.

> **(3)** does this work with _unsigned_ txs? if so, how do you handle vsizes & fee rates etc?

Currently there is no check in the code for if an input is signed or not, meaning an unsigned transaction shows just like a signed one but without signatures (vsizes, fee rates and sigops being the ones of the transaction without the signatures).

Currently we don't check if an input is signed or not, meaning an unsigned transaction appears identical to a signed one, except inputs are not signed. vsizes, fee rates, and sigops reflect the values of the transaction without the signatures.

| Unsigned  | Signed |
| ------------- | ------------- |
| <img width="847" alt="Screenshot 2024-11-28 at 10 39 24" src="https://github.com/user-attachments/assets/cad228af-803d-4bee-81c6-7f6b907bf7ee"> | <img width="847" alt="Screenshot 2024-11-28 at 10 38 59" src="https://github.com/user-attachments/assets/6c611be1-3d76-46bf-b3d8-cbb4b0327cac">  |

The only difference is seen when broadcasting, which obviously fails for the unsigned transaction:
<img width="750" alt="Screenshot 2024-11-28 at 10 37 12" src="https://github.com/user-attachments/assets/2b44a527-f293-4c4f-b0b6-9b0308413beb">

It would be cool to check if inputs are signed, show some indication on inputs that are not, and adapt fee rate and other metrics to include signed data, since the fee rate of an unsigned transaction is useless and can be misleadingly low (however legacy scripts should be handled carefully because the signature changes the txid?)

> **(4)** it would be neat to support PSBTs. that's definitely a job for a future PR, but just mentioning it now in case it makes a difference to the design.

I'll look into this yeah that's a great idea.

softsimon commented:
This error `This transaction is missing data` looks like it's a problem with your own transaction.

Should say something else like Couldn't load Prevouts or something

mononaut commented:
> Sure, will do! some of the code for decoding raw transactions has been written by you in the mempool backend. I also adapted some logic from the bitcoinjs repo.

great, thanks! yeah I recognized some of the bitcoinjs code, but wasn't sure where that ended and your own stuff began.

mononaut commented:
> I agree that an endpoint for directly fetching prevouts instead of handling them client-side would be nice. And a batched prevout endpoint in `mempool/electrs` would be neat! The process for querying a prevout would be:
> 
> 1. See if the prevout is unconfirmed in-memory
> 2. If not, call `gettxout` with `include_mempool=false`
> 
> IIUC this should always return the prevout if the corresponding output exists and is unspent, and also provide extra data for CPFP calculation if it's unconfirmed.

yup, that sounds about right

mononaut commented:
re. unsigned transactions:

yeah I think we definitely need to check if the transaction is fully signed, and otherwise indicate somehow that all of the vsize-based fields are only provisional.

you could add the most likely witness data sizes for common scripts/address formats to the provisional vsize to make that more accurate - maybe check some wallet implementations for inspiration, since estimating final fee rates before signatures are added is a common problem.

if we know a transaction is not fully signed we can also preemptively disable the broadcast button.

wiz commented:
@mononaut is this ok to merge?

natsoni commented:
I need to handle the fetching of prevouts differently, will push a fix soon and I think it will be shippable.

natsoni commented:
We now fetch transaction's prevouts from mempool backend and/or Core, according to this logic: 

> 1. See if the prevout is unconfirmed in-memory
> 2. If not, call `gettxout` with `include_mempool=false`

To do next: 
- about unsigned transactions: 
   - check if each input is signed, and show some indication if unsigned
   - if missing signatures, estimate final vsize-based fields by adding the most likely witness data sizes for common scripts/address formats
   - if we can't do the above because of e.g. unknown input format, indicate that all of the vsize-based fields are only provisional
- compute effective fee rate using CPFP data, if any
- psbt support

natsoni commented:
It’s probably easier to review/merge at this point and go on with following points in another PR, because they will add more stuff on top on what’s already here.

natsoni commented:
_Note: I didn't implement the following in another PR because it changes how some things are handled in this PR, sorry 😅_

Now effective fee rate, if any, is calculated on the decoded transaction page. At first I looked into how to compute cpfp clusters using the prevouts’ unconfirmed transactions client-side, but it felt cumbersome and it’s not accurate in many cases.

Commit 2987f86 creates a new endpoint `POST /api/v1/cpfp` that computes the potential CPFP tree of the decoded transaction and returns its effective fee rate and ancestors. We only query this endpoint if at least one input is spending from an unconfirmed transaction (note that the frontend doesn't need the full unconfirmed transaction anymore because everything is done in the backend).

Also, commit d852c48 moves the CPFP table to its own component to be reused in the decoded transaction component.

mononaut reviewed (approved):
Tested ACK @ [74fa3c7e]

(I made some minor changes to the error handling to conform to the new standard we adopted after this PR was first opened)

I tested this with a variety of transaction styles and it works great.

A few minor inline comments below, but these can probably be addressed in a later PR.

natsoni commented on backend/src/api/bitcoin/bitcoin.routes.ts:
```diff
@@ -930,6 +932,80 @@ class BitcoinRoutes {
+    } catch (e) {
+      handleError(req, res, 500, e instanceof Error ? e.message : e);
+    }
+  }
```
Done in 0a5ce1d, this API now calculates the CPFP clusters of up to 25 transactions. Note that to do so, we temporarily add these transactions to the mempool cache, and remove them once CPFP clusters are calculated.

mononaut replied:
temporarily modifying any of the global mempool data structures makes me very nervous. I get why it simplifies the implementation here, but I can't help think it's going to bite us in the ass later.

touching the spendMap seems particularly dangerous because the add and remove operations aren't symmetric (e.g. if one of the txs conflicts with an existing spend in our mempool, it'll clobber that entry in the spendMap when we add it, but won't restore the previous entry at the end)

I'd prefer to keep the original implementation for now in this PR at least:
 - throw a ">1 txs not yet supported" error if the user passes more than one transaction.
 - otherwise unwrap the single element in the array and proceed as we did before.

then we can do a separate, more focused PR later to implement multi-tx support safely.

natsoni replied:
Right, now that you bring that it seems quite dangerous to do so 😅 will revert these changes

natsoni replied:
> * throw a ">1 txs not yet supported" error if the user passes more than one transaction.
> * otherwise unwrap the single element in the array and proceed as we did before.

Done in 6c95cd2

natsoni commented on backend/src/api/bitcoin/bitcoin.routes.ts:
```diff
@@ -930,6 +932,80 @@ class BitcoinRoutes {
+            unconfirmed = true;  
+          }
+        } else {
+          const rawPrevout = await bitcoinClient.getTxOut(outpoint.txid, outpoint.vout, false);
```
Done in af0c78b

wiz commented:
If this new preview functionality is going to replace the existing push TX page we should probably make it more clear to the user that the transaction has not been broadcast yet - just adding the word "preview" is probably not enough

natsoni commented:
I don’t think this preview page will replace the push transaction page because it’s nice to have the option to directly push a transaction from hex (without previewing it). But I agree that we should make it more explicit on the preview that the transaction hasn’t been broadcast.

mononaut commented:
maybe the screen where you enter the raw hex could just have two separate buttons for "broadcast" and "preview"?

mononaut commented:
Tested re-ACK @ [860bc7d1]

natsoni commented:
How about adding a simple explanation: 
<img width="847" alt="Screenshot 2025-01-14 at 10 26 11" src="https://github.com/user-attachments/assets/b23d311b-8d22-4360-a851-354d039c8172" />

nymkappa commented:
> How about adding a simple explanation: <img alt="Screenshot 2025-01-14 at 10 26 11" width="847" src="https://private-user-images.githubusercontent.com/46578910/402761703-b23d311b-8d22-4360-a851-354d039c8172.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgzMTAwMzAsIm5iZiI6MTczODMwOTczMCwicGF0aCI6Ii80NjU3ODkxMC80MDI3NjE3MDMtYjIzZDMxMWItOGQyMi00MzYwLWE4NTEtMzU0ZDAzOWM4MTcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMxVDA3NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMjFiY2I2MjYxNTRiZmZmM2JmYmE1MTM0YmYwODc1ZDM2ZDgxMTM1NDYwMjA0YmFlNGIxNTQxMTAwNDMyODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BN-1D6UFT_yFsa3MWCK2zz-amwtyXD8qHJ6ZjAjE6AE">

In this case there should not be any ETA since it's not broadcasted yet. Add the CTA to the `Broadcast` work in the alert banner instead of having a button at the top right?

natsoni commented:
> In this case there should not be any ETA since it's not broadcasted yet. Add the CTA to the `Broadcast` work in the alert banner instead of having a button at the top right?

Removed the ETA from the preview and moved the broadcast button to the alert banner: 

https://github.com/user-attachments/assets/a7f927a5-6298-486b-a143-1290162a8da8

softsimon reviewed (approved):
partially tested ACK @ [9a4b5fda6]
